source/backtraceMetric.ts

```diff
@@ -0,0 +1,226 @@
+      this.createNewSession();
+      this.sendUniqueEvent(this.sessionId);
+      // can also send "session length" by sending lastActive - sessionStart.
+      // This can get a sense of "session length" / "time spent on app", "session free minutes", "avg time per error" etc
+    }
+
+    // handle summed events
+    // ex. page load / event
+  }
+
+  /**
+   * Send POST to unique-events API endpoint
+   */
+  private async sendUniqueEvent(sessionId: string): Promise<void> {
+    const endpoint = `https://${this.DIRECTORY}.backtrace.io/api/unique-events/submit?universe=${this.universe}&token=${this.token}`;
+
+    const payload = {
+      application: this.applicationName,
+      appversion: __VERSION__,
+      metadata: {
+        dropped_events: 0,
+      },
+      unique_events: [
+        {
+          timestamp: this.CURRENT_TIMESTAMP,
+          unique: sessionId,
+          attributes: {
+            guid: sessionId,
+            'uname.sysname': this.userAgent,
+            'application.version': __VERSION__,
+            'application.session': '', // application session ?
+          },
+        },
+      ],
+    };
+
+    await this.post(endpoint, JSON.stringify(payload));
+    return;
+  }
+
+  /**
+   * Send POST to summed-events API endpoint
+   */
+  private async sendSummedEvent(
+    sessionId: string,
+    metricGroup: string,
+  ): Promise<void> {
+    const endpoint = `https://${this.DIRECTORY}.backtrace.io/api/summed-events/submit?universe=${this.universe}&token=${this.token}`;
+
+    const payload = {
+      application: this.applicationName,
+      appversion: __VERSION__,
+      metadata: {
+        dropped_events: 0,
+      },
+      summed_events: [
+        {
+          timestamp: this.CURRENT_TIMESTAMP,
+          metric_group: metricGroup,
+          attributes: {
+            guid: sessionId,
+            'uname.sysname': this.userAgent,
+            'application.version': __VERSION__,
+            'application.session': '', // application session ?
+          },
+        },
+      ],
+    };
+
+    await this.post(endpoint, JSON.stringify(payload));
+    return;
+  }
+
+  /**
+   * Send POST request.
+   * @param url - string endpoint
+   * @param data - stringified JSON object
+   */
+  private async post(url: string, data: string): Promise<void> {
+    try {
+      return new Promise<void>((res, rej) => {
+        const http = new XMLHttpRequest();
+        http.setRequestHeader('Content-type', 'application/json');
+        http.timeout = this.timeout;
+        http.open('POST', url, true);
+        http.send(data);
+        http.onload = (e) => {
+          if (http.readyState === XMLHttpRequest.DONE) {
+            if (http.status === 200) {
+              res();
+            } else if (http.status === 429) {
+              rej('Backtrace - reached metric limit.');
+            } else {
+              rej(
+                `Invalid attempt to submit metric to Backtrace. Result: ${http.responseText}`,
```

Comment:
It's worth to also add a status code - based on the status code a lot of people can learn why exactly something happened - for example 404/401/404 vs 501/505 etc.

Reply:
Added http code to error